### PR TITLE
aligned C implementations for memcpy, memset and memmove

### PR DESCRIPTION
**Pull request by teknoraver**
Please give it a try

### CODE REVIEW

File: arch/riscv/lib/string.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+ */
+void *memcpy(void *dest, const void *src, size_t count)
+{
+	static const void *labels[] = {
+		&&u64, &&u8, &&u16, &&u8,
+		&&u32, &&u8, &&u16, &&u8,
+	};
+	union types d = { .u8 = dest };
+	union ctypes s = { .u8 = src };
+#ifdef HAVE_EFFICIENT_UNALIGNED_ACCESS
```
Comment (esmil) [inline arch/riscv/lib/string.c:51]: This should be
```
#ifdef CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS
```
right?

File: arch/riscv/lib/string.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ * Do not use memset() to access IO space, use memset_io() instead.
+ */
+void *memset(void *s, int c, size_t count)
+{
+	const int bytes_long = BITS_PER_LONG / 8;
+	u8 cc[] = { [0 ... bytes_long-1] = c };
+	union ctypes src = { .u8 = cc };
+	union types dest = { .u8 = s };
+
+#ifndef HAVE_EFFICIENT_UNALIGNED_ACCESS
```
Comment (esmil) [inline arch/riscv/lib/string.c:136]: Ditto

File: arch/riscv/lib/string.c (hunk 1 above)
Comment (teknoraver) [reply arch/riscv/lib/string.c:51]: Right!

File: arch/riscv/lib/string.c (hunk 2 above)
Comment (teknoraver) [reply arch/riscv/lib/string.c:136]: So it turns out that memcpy() was working correctly, while memset() was generating unaligned accesses.
I daw memset on top of `perf top` but I couldn't explain why.
Thanks!

Comment (teknoraver) [reply arch/riscv/lib/string.c:136]: Well no, that shouldn't have effect anyway.
I see memset always being called when transmitting small packets, maybe it's normal

**Comment by teknoraver** [comment]
@esmil I wrote a new C memcpy() which always copies 64 bit at time. This time the impact is very noticeable even in perf:

## before
```
$ iperf3 -c beaglev
Connecting to host beaglev, port 5201
[  5] local 192.168.85.6 port 44840 connected to 192.168.85.48 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  76.4 MBytes   641 Mbits/sec   27    624 KBytes
[  5]   1.00-2.00   sec  72.5 MBytes   608 Mbits/sec    0    708 KBytes
[  5]   2.00-3.00   sec  73.8 MBytes   619 Mbits/sec   10    451 KBytes
[  5]   3.00-4.00   sec  72.5 MBytes   608 Mbits/sec    0    564 KBytes
[  5]   4.00-5.00   sec  73.8 MBytes   619 Mbits/sec    0    658 KBytes
[  5]   5.00-6.00   sec  73.8 MBytes   619 Mbits/sec   14    522 KBytes
[  5]   6.00-7.00   sec  73.8 MBytes   619 Mbits/sec    0    621 KBytes
[  5]   7.00-8.00   sec  72.5 MBytes   608 Mbits/sec    0    706 KBytes
[  5]   8.00-9.00   sec  73.8 MBytes   619 Mbits/sec   20    580 KBytes
[  5]   9.00-10.00  sec  73.8 MBytes   619 Mbits/sec    0    672 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   736 MBytes   618 Mbits/sec   71             sender
[  5]   0.00-10.01  sec   733 MBytes   615 Mbits/sec                  receiver

iperf Done.

Overhead  Shared O  Symbol
  42.22%  [kernel]  [k] memcpy
  35.00%  [kernel]  [k] __asm_copy_to_user
   3.50%  [kernel]  [k] sifive_l2_flush64_range
   2.30%  [kernel]  [k] stmmac_napi_poll_rx
   1.11%  [kernel]  [k] memset
```
## after:
```
$ iperf3 -c beaglev
Connecting to host beaglev, port 5201
[  5] local 192.168.85.6 port 44864 connected to 192.168.85.48 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   109 MBytes   912 Mbits/sec   48    559 KBytes
[  5]   1.00-2.00   sec   108 MBytes   902 Mbits/sec    0    690 KBytes
[  5]   2.00-3.00   sec   106 MBytes   891 Mbits/sec   36    396 KBytes
[  5]   3.00-4.00   sec   108 MBytes   902 Mbits/sec    0    567 KBytes
[  5]   4.00-5.00   sec   106 MBytes   891 Mbits/sec    0    699 KBytes
[  5]   5.00-6.00   sec   106 MBytes   891 Mbits/sec   32    414 KBytes
[  5]   6.00-7.00   sec   106 MBytes   891 Mbits/sec    0    583 KBytes
[  5]   7.00-8.00   sec   106 MBytes   891 Mbits/sec    0    708 KBytes
[  5]   8.00-9.00   sec   106 MBytes   891 Mbits/sec   28    433 KBytes
[  5]   9.00-10.00  sec   108 MBytes   902 Mbits/sec    0    591 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.04 GBytes   897 Mbits/sec  144             sender
[  5]   0.00-10.01  sec  1.04 GBytes   894 Mbits/sec                  receiver

iperf Done.

Overhead  Shared O  Symbol
  45.69%  [kernel]  [k] __asm_copy_to_user
  29.06%  [kernel]  [k] memcpy
   4.09%  [kernel]  [k] sifive_l2_flush64_range
   2.77%  [kernel]  [k] stmmac_napi_poll_rx
   1.24%  [kernel]  [k] memset
```

**Pull request closed without merging**